### PR TITLE
WIP cli-caching [skip-ci]

### LINKED ISSUE #1300: Proposal: download Cypress to the common cached folder
## Feature

Instead of downloading and unzipping Cypress into `./node_modules/cypress/dist/` download it into a global folder for better caching. This will allow

- better caching of dependencies on CI
- reusing same version of Cypress if already downloaded once for some project

To determine output location use https://github.com/LinusU/node-cachedir

```js
var cachedir = require('cachedir')
var path = cachedir('cypress')
// `path` now contains the path under which you should store cached files
```

- cypress run, open, verify
  - [ ] ~take arg `--binary-folder` (looks like: `/foo/bar/Cypress.app` (macOS) or `/foo/bar/Cypress` (linux, win), to override default path to cypress binary folder (default is  `(cypressCache)/<cli version>/Cypress.app`)~
  - [x] can override folder of binary with env var `CYPRESS_BINARY_FOLDER=/path/to/Cypress.app`
    - [x] warn user if set

- cypress install
  - [x] Checks `(cypressCache)/<cli version>` for already installed binary folder (`Cypress.app` on macOS, `Cypress` on linux/win), and if binary version matches `(installVersion)`, do nothing.
  - [x] `(installVersion)` defaults to cli version, is overridden by env var `CYPRESS_BINARY_VERSION` during install
    - can be either `semvar: 1.2.3`, `url: www.cypress.io/download/1.2.3`, or `zip: /path/to/cypress.zip`
  - [x] if `cypress install --force`, force reinstall to `(cypressCache)/<cli version>`
  - [x] `(cypressCache)` defaults to:
    - `~/Library/Caches/Cypress` (macOS)
    - `~/.cache/Cypress` (linux)
    - `/AppData/Local/Cypress/Cache` (win)
  -  [x] `(cypressCache)` can be overridden with env var `CYPRESS_CACHE_DIRECTORY`
  - [x] before install, ensure write access to `(cypressCache)`

- [x] custom error message for failed access to `(cypressCache)`
- [ ] docs
  - [x] Installing Cypress
  - [ ] ~command line usage~
  - [ ] update info about global install
- [x] manual testing all use cases
  
  - [x] cypress install
    - CYPRESS_BINARY_VERSION
      - [x] url 
        - [x] skip if has any cache
      - [x] semvar
      - [x] zip
        - [x] skip if has any cache
        - [x] show progress for unzipping 
      - [x] --force
    - [x] --force
  - [x] ~--binary-folder~
    - [x] ~cypress run~
    - [x] ~cypress open~
    - [x] ~cypress verify~
    - [x] ~invalid~
  - [x] CYPRESS_CACHE_DIRECTORY
    - valid
    - invalid, error message
  

- [x] add local code coverage with `nyc`
  - [ ] ~~send to `coveralls` in CI~~
## Stretch goal
- [ ] can we cache `ffmpeg` separately to avoid bundling the same version over and over?

### PR DESCRIPTION
**Pull request by Bkucera**
closes #1300 #1254 #890

### CODE REVIEW

**Comment by KellyR-STCU** [comment]
Pretty pumped about this. Thanks for the work

File: cli/lib/tasks/info.js [hunk 1]
```diff
@@ -16,44 +18,77 @@ const getPlatformExecutable = () => {
   }
 }
 
-const getInstallationDir = () => {
+const getInstallationDir = (version = util.pkgVersion()) => {
+  return getInfoFileContents()
+  .tap(debug)
+  .get('install_directory')
+  .catch(() => {
+    let cache_directory = cachedir('Cypress')
```
Comment (bahmutov) [inline cli/lib/tasks/info.js]: should we use capital letter? I would prefer lower case for simplicity

File: cli/lib/tasks/info.js [hunk 2]
```diff
@@ -16,44 +18,77 @@ const getPlatformExecutable = () => {
...
+
+    if (process.env.CYPRESS_CACHE_DIRECTORY) {
```
Comment (bahmutov) [inline cli/lib/tasks/info.js]: also, this entire function could be just a ternary
```
() => {
  const dir = process.env.CYPRESS_CACHE_DIRECTORY 
    ? process.env.CYPRESS_CACHE_DIRECTORY : cachedir('Cypress')
  return path.join(dir, version)
}
```

File: cli/lib/tasks/install.js [hunk 3]
```diff
@@ -14,6 +14,7 @@ const unzip = require('./unzip')
 const logger = require('../logger')
 const la = require('lazy-ass')
 const is = require('check-more-types')
+// const cachedir = require('cachedir')
```
Comment (bahmutov) [inline cli/lib/tasks/install.js]: we usually do not leave commented out code, because it is 
- just confusing
- not tested

**Comment by bahmutov** [comment]
love it

**Comment by brian-mann** [comment]
@bahmutov why did you close this PR?

I told @Bkucera to open a WIP so we could indicate that this is being worked on.

File: cli/lib/tasks/info.js (hunk 1 above)
Comment (brian-mann) [reply cli/lib/tasks/info.js]: I believe its canonical to capitalize the name. Looking through my `~/Library/Application Support` and `/Library/Caches` most apps (including Electron) have a capitalized folder containing the cached + app data.

Comment (brian-mann) [reply cli/lib/tasks/info.js]: screen shot 2018-04-14 at 3 00 39 pm

**Comment by bahmutov** [comment]
Oh my, sure, I just wanted to avoid case sensitive things, and reuse package json name

**Comment by Bkucera** [comment]
small decision: info.json in node_modules/cypress/dist/info.json will include `version` and `cache_directory`, so the binary can be found are they going to be set before or after download?
I vote after download so if you kill `cypress install`, the info.json will not have been written to yet

**Comment by brian-mann** [comment]
Yes - state should only be set once everything has completed.

**Comment by Bkucera** [comment]
if cache is deleted, and cypress is ran, should it alert the user to run `cypress install`, or automatically do it?

**Comment by bahmutov** [comment]
Exactly! User should know most importantly what to do and how to fix this problem. Even consulting our docs is ok, but some action is a must

**Pull request closed without merging**